File: resources/android/SwitchRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme

sealed interface SwitchRendererEvent {
    data class ToggleChange(
        val callbackId: Int,
        val nodeId: Int,
        val value: Boolean,
    ) : SwitchRendererEvent
}

data class SwitchRendererConfiguration(
    val nodeId: Int,
    val value: Boolean,
    val label: String,
    val helper: String,
    val error: String,
    val disabled: Boolean,
    val accessibilityLabel: String,
    val accessibilityHint: String,
    val onChangeCallback: Int,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        value = node.props.getBool("value"),
        label = node.props.getString("label"),
        helper = node.props.getString("helper"),
        error = node.props.getString("error"),
        disabled = node.props.getBool("disabled"),
        accessibilityLabel = node.props.getString("a11y_label").ifEmpty {
            node.props.getString("label")
        },
        accessibilityHint = node.props.getString("a11y_hint"),
        onChangeCallback = node.props.getCallbackId("on_change"),
    )

    val supportingText: String
        get() = error.ifEmpty { helper }
}

class SwitchRendererState(node: NativeUINode) {
    var configuration by mutableStateOf(SwitchRendererConfiguration(node))
        private set

    private var pendingProposal: Boolean? = null

    fun proposeChange(): SwitchRendererEvent.ToggleChange? {
        if (configuration.disabled || configuration.onChangeCallback == 0 || pendingProposal != null) {
            return null
        }

        val proposal = !configuration.value
        pendingProposal = proposal

        return SwitchRendererEvent.ToggleChange(
            callbackId = configuration.onChangeCallback,
            nodeId = configuration.nodeId,
            value = proposal,
        )
    }

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findSwitchNode(configuration.nodeId) ?: return false
        val previousValue = configuration.value
        configuration = SwitchRendererConfiguration(node)
        pendingProposal = null
        return previousValue != configuration.value
    }
}

object SwitchRenderer {
    @Composable
    fun Render(node: NativeUINode, modifier: Modifier) {
        val state = remember(node.id) { SwitchRendererState(node) }
        val publishedTree = NativeUIBridge.currentTree.value
        LaunchedEffect(publishedTree) {
            publishedTree?.let(state::serverPublished)
        }

        val configuration = state.configuration
        val tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light

        FirstlightSwitchControl(
            value = configuration.value,
            label = configuration.label,
            helper = configuration.helper,
            error = configuration.error,
            disabled = configuration.disabled,
            accessibilityLabel = configuration.accessibilityLabel,
            accessibilityHint = configuration.accessibilityHint,
            tokens = tokens,
            onProposal = {
                state.proposeChange()?.let { event ->
                    NativeUIBridge.sendToggleChangeEvent(
                        event.callbackId,
                        event.nodeId,
                        event.value,
                    )
                }
            },
            modifier = modifier,
        )
    }
}

private fun NativeUINode.findSwitchNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findSwitchNode(id)?.let { return it } }
    return null
}

File: resources/ios/SwitchRenderer.swift
import SwiftUI

enum SwitchRendererEvent: Equatable {
    case toggleChange(callbackId: Int, nodeId: Int, value: Bool)
}

struct SwitchRendererConfiguration: Equatable {
    let nodeID: Int
    let value: Bool
    let label: String
    let helper: String
    let error: String
    let disabled: Bool
    let accessibilityLabel: String
    let accessibilityHint: String
    let onChangeCallback: Int

    init(node: NativeUINode) {
        let props = node.props
        let explicitAccessibilityLabel = props.getString("a11y_label")

        nodeID = node.id
        value = props.getBool("value")
        label = props.getString("label")
        helper = props.getString("helper")
        error = props.getString("error")
        disabled = props.getBool("disabled")
        accessibilityLabel = explicitAccessibilityLabel.isEmpty ? label : explicitAccessibilityLabel
        accessibilityHint = props.getString("a11y_hint")
        onChangeCallback = props.getCallbackId("on_change")
    }

    var supportingText: String {
        error.isEmpty ? helper : error
    }
}

struct SwitchRendererState: Equatable {
    var configuration: SwitchRendererConfiguration
    private var pendingProposal: Bool?

    init(node: NativeUINode) {
        configuration = SwitchRendererConfiguration(node: node)
        pendingProposal = nil
    }

    mutating func proposeChange() -> SwitchRendererEvent? {
        guard !configuration.disabled,
              configuration.onChangeCallback != 0,
              pendingProposal == nil
        else {
            return nil
        }

        let proposal = !configuration.value
        pendingProposal = proposal

        return .toggleChange(
            callbackId: configuration.onChangeCallback,
            nodeId: configuration.nodeID,
            value: proposal
        )
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = Self.findNode(id: configuration.nodeID, in: tree.root) else {
            return false
        }

        let previousValue = configuration.value
        configuration = SwitchRendererConfiguration(node: node)
        pendingProposal = nil

        return previousValue != configuration.value
    }

    private static func findNode(id: Int, in node: NativeUINode) -> NativeUINode? {
        if node.id == id {
            return node
        }

        for child in node.children {
            if let match = findNode(id: id, in: child) {
                return match
            }
        }

        return nil
    }
}

struct SwitchRenderer: View {
    @ObservedObject private var themeStore = NativeUITheme.shared
    @ObservedObject private var bridge = NativeUIBridge.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var rendererState: SwitchRendererState

    init(node: NativeUINode) {
        _rendererState = State(initialValue: SwitchRendererState(node: node))
    }

    var body: some View {
        let configuration = rendererState.configuration
        let tokens = themeStore.resolve(for: colorScheme)

        FirstlightSwitchControl(
            value: configuration.value,
            label: configuration.label,
            supportingText: configuration.supportingText,
            error: configuration.error,
            disabled: configuration.disabled,
            accessibilityLabel: configuration.accessibilityLabel,
            accessibilityHint: configuration.accessibilityHint,
            tokens: tokens,
            onProposal: proposeChange
        )
        .onReceive(bridge.$currentTree.compactMap { $0 }) { tree in
            // A PHP response may keep the same accepted value. Reconcile every
            // publication so that response still clears the pending proposal.
            rendererState.serverPublished(tree)
        }
    }

    private func proposeChange() {
        guard case let .toggleChange(callbackId, nodeId, value)? = rendererState.proposeChange() else {
            return
        }

        NativeElementBridge.sendToggleChangeEvent(
            callbackId,
            nodeId: nodeId,
            value: value
        )
    }
}
